### PR TITLE
feat: [MR-551] Factor input scheduling into separate struct

### PR DESCRIPTION
**Pull request by alin-at-dfinity**
[MR-551]: Encapsulate input sender schedules and `next_input_source` within a separate `InputSchedule` struct. Along with simplifying the `CanisterQueues` logic, this provides better separation of concerns and safer invariant checks.

[MR-551]: https://dfinity.atlassian.net/browse/MR-551?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by alin-at-dfinity** [comment]
This looks like a huge change, but apart from the renamed fields and types, it consists of 2 large but not particularly interesting chunks:

* Code moved / copied from `canister_queues.rs` into `input_schedule.rs`: there is no change in logic, just copy-pasta and replacing the `own_canister_id`, `local_canisters` pair with a function pointer (for determining whether a canister is local or not).
* A whole lot of tests for the new `InputSchedule`. There are a handful of edge cases in there (e.g. what happens when the same canister is enqueued again), identified by inline comments.

File: rs/replicated_state/src/canister_state/queues/input_schedule.rs [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+/// we rotate round-robin between the three sources. We also rotate round-robin
+/// over the canisters in the local sender and remote serder buckets when we pop
+/// or skip messages from those sources.
+#[derive(Clone, Debug, Default, PartialEq, Eq, ValidateEq)]
+pub(super) struct InputSchedule {
+    /// The input source (local senders, ingress or remote senders) at the front
+    /// of the schedule.
+    ///
+    /// Used to implement round-robin rotation over input sources.
+    #[validate_eq(Ignore)]
```
Comment (derlerd-dfinity) [inline rs/replicated_state/src/canister_state/queues/input_schedule.rs:33]: Why do we have this as `Ignore`?

File: rs/replicated_state/src/canister_state/queues/input_schedule.rs [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+}
+
+impl InputSchedule {
+    /// Returns the current input source, without advancing it (e.g. for `peek()`).
+    pub(super) fn input_source(&self) -> InputSource {
+        self.next_input_source
+    }
+
+    /// Returns the current input source and advances to the next one.
+    pub(super) fn next_input_source(&mut self) -> InputSource {
```
Comment (derlerd-dfinity) [inline rs/replicated_state/src/canister_state/queues/input_schedule.rs:73]: As far as I can tell these functions are used in tests only at the moment. Is this intended?

**Review by derlerd-dfinity — COMMENTED** [review]
Thank you. Left some questions after an initial pass.

File: rs/replicated_state/src/canister_state/queues/input_schedule.rs (hunk 1 above)
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/queues/input_schedule.rs:33]: No idea, I wanted to ask the same question when I first saw it.

@pakhomov-dfinity?

File: rs/replicated_state/src/canister_state/queues/input_schedule.rs (hunk 2 above)
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/queues/input_schedule.rs:73]: Yes, I didn't want to rewrite the old `CanisterQueues` to make use of this, only to then immediately throw away the new code in #879.

I did rebase #879 on top of this, though. So the new `CanisterQueues` will be relying on this instead of rolling its own input schedule implementation.

**Review by derlerd-dfinity — APPROVED** [review]
Thanks a lot. LGTM.

**Review by oggy-dfin — COMMENTED** [review]
I had a quick look and didn't spot any issues. But I'll only have time for a thorough review tomorrow. But feel free to go ahead and merge meanwhile, since David's already done a review.

File: rs/replicated_state/src/canister_state/queues/input_schedule.rs [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+/// There are three possible outcomes of an inconsistent `InputSchedule`:
+///
+///  * An input queue is scheduled multiple times, leading to unfairness.
+///  * An input queue is never scheduled, leading to starvation.
+///  * Replica divergence, if a restarted replica loads an internally consistent
+///    `InputSchedule`, while other replicas continue with a mismatch between
+///    the two schedules on the one hand and `scheduled_senders` on the other.
+///
+/// All of the above occurrences would be detected from critical errors or other
+/// alerting. Thay are also exceedingly unlikely to result in a crash loop,
```
[the remark this answers is not in the document]
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/queues/input_schedule.rs]: Applied, thanks.

File: rs/replicated_state/src/canister_state/queues/input_schedule/tests.rs [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+    assert_eq!(InputSource::Ingress, schedule.input_source());
+    assert_eq!(InputSource::Ingress, schedule.next_input_source());
+
+    assert_eq!(InputSource::RemoteSubnet, schedule.input_source());
+    assert_eq!(InputSource::RemoteSubnet, schedule.next_input_source());
+
+    assert_eq!(InputSchedule::default(), schedule);
+
+    assert_eq!(InputSource::LocalSubnet, schedule.input_source());
+    assert_eq!(InputSource::LocalSubnet, schedule.next_input_source());
```
[the remark this answers is not in the document]
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/queues/input_schedule/tests.rs:25]: We just tested that it circled back to `LocalSubnet` (`InputSource::default()` is `LocalSubnet`; which is also why we're checking that the schedule as a whole is equal to `InputSchedule::default()`).

We would just be doing N+2 iterations instead of N+1.

File: rs/replicated_state/src/canister_state/queues/input_schedule/tests.rs [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+            LocalSubnet => RemoteSubnet,
+            RemoteSubnet => LocalSubnet,
+        },
+    );
+}
+
+#[test]
+#[should_panic]
+fn test_reschedule_in_remote_while_in_local() {
+    test_reschedule_while_in_other_queue(LocalSubnet);
```
[the remark this answers is not in the document]
Comment (alin-at-dfinity) [reply rs/replicated_state/src/canister_state/queues/input_schedule/tests.rs:183]: Precisely.

Added some inline comments to `test_reschedule_while_in_other_queue()` above.
